### DISCUSSION TITLE
SBOM scan fails on package name with trailing "/"

### DISCUSSION THREAD

**Discussion by Bah-bah** [aquasecurity/trivy#6259] [Bugs]
### Description

Hi!
I found this image on dockerhub:
`docker pull bionode/cocalc-kubernetes-project`
I scan it in two steps: first, create SBOM, then scan it.
```
OUTPUT=$(docker run -v /dir/app aquasec/trivy rootfs app/bionode-cocalc-kubernetes-project-fs/ --format cyclonedx --scanners vuln -q)
echo "$OUTPUT" > image.cyclonedx
OUTPUT=$(docker run -v /dir:/app aquasec/trivy sbom app/image.cyclonedx --format json -q)
echo "$OUTPUT" > image-rootfs.json
```
SBOM is created successfully, but scan fails with
`FATAL	sbom scan error: scan error: scan failed: failed analysis: SBOM decode error: failed to decode: failed to parse sbom: failed to aggregate packages: failed to parse the component: failed to parse purl: failed to parse purl(pkg:npm/event-stream_tests/@0.0.0): purl is missing name`
I looked inside SBOM and found that `event-stream_tests/` is the only component there that has trailing slash. Component itself:
```
    {
      "bom-ref": "pkg:npm/event-stream_tests/@0.0.0?file_path=cocalc%2Fsrc%2Fnode_modules%2Fevent-stream%2Ftest%2Fpackage.json",
      "type": "library",
      "name": "event-stream_tests/",
      "version": "0.0.0",
      "purl": "pkg:npm/event-stream_tests/@0.0.0",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgID",
          "value": "event-stream_tests/@0.0.0"
        },
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "node-pkg"
        },
        {
          "name": "aquasecurity:trivy:FilePath",
          "value": "cocalc/src/node_modules/event-stream/test/package.json"
        }
      ]
    }
```
If I remove slash from its name scan ends successfully. It makes me think that this kind of name - with trailing slash - is considered malformed.
I don't know much about node.js and don't understand npm's build process, thus cannot say what is wrong with this package. But from my perspective either SBOM creation should make other name for it or scanning of SBOM should not fail on it.

### Desired Behavior

SBOM doesn't include component with such name
**OR**
scan shouldn't fail on it

### Actual Behavior

Scan fails with failed to parse `purl(pkg:npm/event-stream_tests/@0.0.0): purl is missing name`

### Reproduction Steps

```bash
1.docker pull bionode/cocalc-kubernetes-project
2.docker run -v /dir/app aquasec/trivy rootfs app/bionode-cocalc-kubernetes-project-fs/ --format cyclonedx --scanners vuln -q > image.cyclonedx
3.docker run -v /dir:/app aquasec/trivy sbom app/image.cyclonedx --format json -q > image-rootfs.json
```

### Target

None

### Scanner

None

### Output Format

None

### Mode

None

### Debug Output

```bash
FATAL	sbom scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.Run
        /home/runner/work/trivy/trivy/pkg/commands/artifact/run.go:441
  - scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.(*runner).scanArtifact
        /home/runner/work/trivy/trivy/pkg/commands/artifact/run.go:269
  - scan failed:
    github.com/aquasecurity/trivy/pkg/commands/artifact.scan
        /home/runner/work/trivy/trivy/pkg/commands/artifact/run.go:706
  - failed analysis:
    github.com/aquasecurity/trivy/pkg/scanner.Scanner.ScanArtifact
        /home/runner/work/trivy/trivy/pkg/scanner/scan.go:148
  - SBOM decode error:
    github.com/aquasecurity/trivy/pkg/fanal/artifact/sbom.Artifact.Inspect
        /home/runner/work/trivy/trivy/pkg/fanal/artifact/sbom/sbom.go:55
  - failed to decode:
    github.com/aquasecurity/trivy/pkg/sbom.Decode
        /home/runner/work/trivy/trivy/pkg/sbom/sbom.go:225
  - failed to parse sbom:
    github.com/aquasecurity/trivy/pkg/sbom/cyclonedx.(*BOM).UnmarshalJSON
        /home/runner/work/trivy/trivy/pkg/sbom/cyclonedx/unmarshal.go:60
  - failed to aggregate packages:
    github.com/aquasecurity/trivy/pkg/sbom/cyclonedx.(*BOM).parseSBOM
        /home/runner/work/trivy/trivy/pkg/sbom/cyclonedx/unmarshal.go:147
  - failed to parse the component:
    github.com/aquasecurity/trivy/pkg/sbom/cyclonedx.aggregatePkgs
        /home/runner/work/trivy/trivy/pkg/sbom/cyclonedx/unmarshal.go:290
  - failed to parse purl:
    github.com/aquasecurity/trivy/pkg/sbom/cyclonedx.toPackage
        /home/runner/work/trivy/trivy/pkg/sbom/cyclonedx/unmarshal.go:347
  - failed to parse purl(pkg:npm/event-stream_tests/@0.0.0):
    github.com/aquasecurity/trivy/pkg/purl.FromString
        /home/runner/work/trivy/trivy/pkg/purl/purl.go:54
  - purl is missing name
```

### Operating System

macOS

### Version

```bash
Version: 0.49.1
```

### Checklist

- [ ] Run `trivy image --reset`
- [ ] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by DmitriyLewen** [comment]
Hello @Bah-bah 
Thanks for your report!

Can you share `cocalc/src/node_modules/event-stream/test/package.json` file from this image?
Looks like this file contains incorrect name.

**Reply by Bah-bah** [reply]
Hi @DmitriyLewen 
Unfortunately, I cannot. It is not my image (it's public). And don't know how to make such npm package :(

**Reply by DmitriyLewen** [reply]
My `docker` doesn't want to pull this image because it is very large 😄

I was able to get this file using the IDE debugger:
```
{ "name": "event-stream_tests/"
, "version": "0.0.0"
, "description": ""
, "homepage": "http://github.com/dominictarr/event-stream_tests/"
, "repository": 
  { "type": "git"
  , "url": "https://github.com/dominictarr/event-stream_tests/.git" }
, "dependencies": {}
, "devDependencies": {}
, "author": "Dominic Tarr <dominic.tarr@gmail.com> (http://bit.ly/dominictarr)"
, "scripts": { "test": "meta-test test/*.js" } }
```

Look on `"name": "event-stream_tests/"`. This name is incorrect -https://docs.npmjs.com/cli/v10/configuring-npm/package-json.

**Reply by Bah-bah** [reply]
Is there a way to ignore such malformed packages on scanning? Or not to include it in SBOM?

**Reply by DmitriyLewen** [reply]
I check `npm: `npm` returns error for this case:
```
➜  npm init
This utility will walk you through creating a package.json file.
It only covers the most common items, and tries to guess sensible defaults.


See `npm help init` for definitive documentation on these fields
and exactly what they do.


Use `npm install <pkg>` afterwards to install a package and
save it as a dependency in the package.json file.


Press ^C at any time to quit.
package name: (6259) testpackage/
Sorry, name can only contain URL-friendly characters.
```

We should try to behave the same way (return error).
I will create PR to fix this.

**Reply by Bah-bah** [reply]
Return error on creation of SBOM or on scanning?

**Reply by DmitriyLewen** [reply]
Return error when scanning `package.json` file.

**Reply by DmitriyLewen** [reply]
created https://github.com/aquasecurity/trivy/pull/6268

**Reply by Bah-bah** [reply]
Does it mean that after this update I won't be able to receive any results from scanning of this SBOM? It's a bit sad. I'd prefer to skip this package and get other results.

**Reply by DmitriyLewen** [reply]
With these changes, we will no longer include this package in SBOM file.
We will show a warning that this `package.json` file is invalid.

You will need to rescan this image to obtain SBOM file. Trivy will not return error for new SBOM file.
We will still return error for old SBOM file because in this case your old SBOM is invalid so we need to return an error.
